revocation/revocation_test.go

```diff
@@ -0,0 +1,582 @@
+	for i, cert := range misorderedIntermediateChain {
+		if i != (len(misorderedIntermediateChain) - 1) {
+			// Skip root which won't have an OCSP Server
+			cert.OCSPServer[0] = fmt.Sprintf("http://example.com/chain_ocsp/%d", i)
+		}
+	}
+
+	missingIntermediateChain := []*x509.Certificate{revokableTuples[0].Cert, revokableTuples[2].Cert, revokableTuples[3].Cert}
+	for i, cert := range missingIntermediateChain {
+		if i != (len(missingIntermediateChain) - 1) {
+			// Skip root which won't have an OCSP Server
+			cert.OCSPServer[0] = fmt.Sprintf("http://example.com/chain_ocsp/%d", i)
+		}
+	}
+
+	invalidChainErr := revocation_ocsp.CheckOCSPError{Err: errors.New("invalid chain: expected chain to be correct and complete with each cert issued by the next in the chain")}
+
+	t.Run("chain missing intermediate", func(t *testing.T) {
+		client := testhelper.MockClient(revokableTuples, []ocsp.ResponseStatus{ocsp.Good}, nil, true)
+		r := New(client)
+
+		errs := r.Validate(missingIntermediateChain, time.Now())
+		expectedErrs := [][]error{
+			{invalidChainErr},
+			{invalidChainErr},
+			{invalidChainErr},
+		}
+		if !identicalErrResults(errs, expectedErrs) {
+			t.Errorf("Expected invalid chain error.\nReceived: %v\nExpected: %v\n", errs, expectedErrs)
+		}
+	})
+
+	t.Run("chain out of order", func(t *testing.T) {
+		client := testhelper.MockClient(misorderedIntermediateTuples, []ocsp.ResponseStatus{ocsp.Good}, nil, true)
+		r := New(client)
+
+		errs := r.Validate(misorderedIntermediateChain, time.Now())
+		expectedErrs := [][]error{
+			{invalidChainErr},
+			{invalidChainErr},
+			{invalidChainErr},
+			{invalidChainErr},
+		}
+		if !identicalErrResults(errs, expectedErrs) {
+			t.Errorf("Expected invalid chain error.\nReceived: %v\nExpected: %v\n", errs, expectedErrs)
+		}
+	})
+}
+
+func TestResultString(t *testing.T) {
+	t.Run("ok", func(t *testing.T) {
+		if OK.String() != "OK" {
+			t.Errorf("Expected %s but got %s", "OK", OK.String())
+		}
+	})
+	t.Run("unknown", func(t *testing.T) {
+		if Unknown.String() != "Unknown" {
+			t.Errorf("Expected %s but got %s", "Unknown", Unknown.String())
+		}
+	})
+	t.Run("revoked", func(t *testing.T) {
+		if Revoked.String() != "Revoked" {
+			t.Errorf("Expected %s but got %s", "Revoked", Revoked.String())
+		}
+	})
+	t.Run("invalid result", func(t *testing.T) {
+		if Result(3).String() != "Invalid Result" {
+			t.Errorf("Expected %s but got %s", "Invalid Result", Result(3).String())
+		}
+	})
+}
+
+func TestGetResultFromErrorsSingle(t *testing.T) {
+	t.Run("nil", func(t *testing.T) {
+		result := GetResultFromErrors([][]error{{nil}})
+		if result != OK {
+			t.Errorf("Expected %s but got %s", OK, result)
+		}
+	})
+	t.Run("no ocsp server", func(t *testing.T) {
+		result := GetResultFromErrors([][]error{{revocation_ocsp.NoOCSPServerError{}}})
+		if result != OK {
+			t.Errorf("Expected %s but got %s", OK, result)
+		}
+	})
```

Comment:
It does not seem right to return `OK` for `NoOCSPServerError`.

Reply:
The thought process here is that we don't consider this an important error. If a certificate in the chain does not support OCSP revocation, that shouldn't cause the chain to fail (since we don't support CRL revocation yet). It might also make sense to change the pkix ocsp no check error to also be OK for this same reason (until CRL is implemented). We could return Unknown, but I think that would be too restrictive in how we are using this in notation-go (where there is no way to differentiate between Unknown and Revoked, both are either logged or enforced)

Reply:
Another alternative could be introducing an additional result type of "MinorError" that would include those two categories, that way it isn't the same as OK but it also isn't grouped into Unknown. We could leverage this by treating MinorError as not a failing error on the notation-go side (while still logging the individual errors as previously described)

Reply:
@priteshbandi What's your comment on this?

Reply:
IMO `GetResultFromErrors` method should be in notation-go not in notation-core-go as this is not a x509 or crypto operation but more of an notation implementation detail.

If a certificate doesnt has OCSP then notation should check for revocation in CRL if CRL is also not present, then we should consider certificate as good because if user hasn't configured OCSP or CRL in there certificate means they dont want revocation support, common example of this is root certificate.